### PR TITLE
Add tool install implicit RIDs

### LINKED ISSUE #21660: Implicit architecture targeting not in place for `dotnet tool install`
The spec talks about [implicit RID targeting](https://github.com/dotnet/designs/blob/main/accepted/2021/x64-emulation-on-arm64/x64-emulation.md#rely-on-implicit-rid-defaults).

From that doc, these are the implicit RIDs:

- .NET 6: osx-arm64
- .NET 5: osx-x64
- .NET Core 3.1: osx-x64

Not documented, on Windows these would be slightly different:

- .NET 6: win-arm64
- .NET 5: win-arm64
- .NET Core 3.1: win-x64

My environment:

```bash
rich@MacBook-Air-M1-2020 app % dotnet --info | grep RID
RID: osx-arm64
rich@MacBook-Air-M1-2020 app % dotnet --version
6.0.100-rc.2.21478.25
```

This is working, AFAICT, for `dotnet run`.

```bash
rich@MacBook-Air-M1-2020 app % cat app.csproj | grep Target
<TargetFramework>net5.0</TargetFramework>
rich@MacBook-Air-M1-2020 app % dotnet run
/Users/rich/app/app.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Host.osx-x64 with version (= 5.0.11)
/Users/rich/app/app.csproj : error NU1102: - Found 2021 version(s) in dotnet6 [ Nearest version: 6.0.0-alpha.1.20423.5 ]
/Users/rich/app/app.csproj : error NU1102: - Found 62 version(s) in nuget.org [ Nearest version: 6.0.0-preview.1.21102.12 ]
/Users/rich/app/app.csproj : error NU1102: - Found 1 version(s) in internal [ Nearest version: 6.0.0-rc.2.21474.18 ]

The build failed. Fix the build errors and run again.
```

That's failing, but due to a mismatch in versions available on the feeds I have registered, so NP. The key thing is that the CLI is asking for the `osx-x64` package, which is the desired one.

Let's look at `dotnet tool install` in contrast.

```bash   
rich@MacBook-Air-M1-2020 app % dotnet tool install -g dotnet-serve
You can invoke the tool using the following command: dotnet-serve
Tool 'dotnet-serve' (version '1.9.71') was successfully installed.
rich@MacBook-Air-M1-2020 app % dotnet-serve 
zsh: killed     dotnet-serve
rich@MacBook-Air-M1-2020 app % codesign -s - ~/.dotnet/tools/dotnet-serve 
rich@MacBook-Air-M1-2020 app % dotnet-serve                              
It was not possible to find any compatible framework version
The framework 'Microsoft.AspNetCore.App', version '5.0.0' (arm64) was not found.
  - The following frameworks were found:
      6.0.0-rc.2.21470.37 at [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
      6.0.0-rc.2.21475.12 at [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]

You can resolve the problem by installing the specified framework and/or SDK.

The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.AspNetCore.App&framework_version=5.0.0&arch=arm64&rid=osx.11.1-arm64
rich@MacBook-Air-M1-2020 app % 
```

It is  pretty obvious that this  isn't a good situation.  The user needs to know a-priori that  `dotnet-serve` needs to be installed with `-a x64`. We  can and should do that automatically, matching `dotnet run` behavior.

### PR DESCRIPTION
**Pull request by sfoslund**
Fixes https://github.com/dotnet/sdk/issues/21660

### CODE REVIEW

**Review by joeloff — APPROVED** [review]

**Review by dsplaisted — APPROVED** [review]
It looks like we don't have any tests for the arm64 / x64 logic this is adding.  Is it possible to add tests for that?

Also check with @richlander that this is still the logic that we want.

**Review by joeloff — APPROVED** [review]

**Comment by sfoslund** [comment]
@dsplaisted I've been seeing the darwin leg timing out repeatedly, is this one of the infrastructure issues you had to fix in the branding PR?

**Comment by dsplaisted** [comment]
> @dsplaisted I've been seeing the darwin leg timing out repeatedly, is this one of the infrastructure issues you had to fix in the branding PR?

No, this doesn't look like the same failure.  The one I hit wasn't even successfully building, where here it looks like the test run is timing out.

File: src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs [hunk 1]
```diff
@@ -32,14 +32,16 @@ internal class ShellShimTemplateFinder
...
             string rid;
             var validRids = new string[] { "win-x64", "win-arm64", "osx-x64", "osx-arm64" };
             if (string.IsNullOrEmpty(archOption))
             {
-                if (!string.IsNullOrEmpty(targetFramework) && new NuGetFramework(targetFramework).Version < new Version("6.0")
-                    && (OperatingSystem.IsWindows() || OperatingSystem.IsMacOS()) && !RuntimeInformation.ProcessArchitecture.Equals(Architecture.X64))
+                if (targetFramework != null &&
+                    ((targetFramework.Version.Major < 6 && OperatingSystem.IsMacOS()) ||
+                    (targetFramework.Version.Major < 5 && OperatingSystem.IsWindows())
+                    && !arch.Equals(Architecture.X64)))
```
[the remark this answers is not in the document]
Comment (sfoslund) [reply src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs:44]: It shouldn't be, Will said that these are the only valid RIDs for these packages: https://github.com/dotnet/sdk/pull/22331/files#diff-88092ceb75a11ccf437cbc0ba4314df63f5eab948e80bc1f21d291b34a92aa70R38

**Pull request merged**